### PR TITLE
Custom Log Formatting

### LINKED ISSUE #3257: Show log location
After switching to slog, log location is gone which is quite useful. Maybe a customized `FullFormat` need to be implemented.

### LINKED ISSUE #3261: log not outputting source
## Bug Report

**What version of Rust are you using?**

**What operating system and CPU are you using?**

**What did you do?**

Run TiKV

**What did you expect to see?**

```
2018/06/22 22:16:24.725 pd.rs:437: [INFO] [region 4128085] try to change peer RemoveNode id: 4604857 store_id: 53
2018/06/22 22:17:23.736 pd.rs:437: [INFO] [region 4128085] try to change peer RemoveNode id: 4604857 store_id: 53
2018/06/22 22:18:22.467 pd.rs:437: [INFO] [region 4128085] try to change peer RemoveNode id: 4604857 store_id: 53
2018/06/22 22:19:20.896 pd.rs:437: [INFO] [region 4128085] try to change peer RemoveNode id: 4604857 store_id: 53
2018/06/22 22:20:19.209 pd.rs:437: [INFO] [region 4128085] try to change peer RemoveNode id: 4604857 store_id: 53
```

**What did you see instead?**

```
Jun 29 20:00:43.749 INFO [region 2] on split region at key t\200\000\000\000\000\000\000\377\027\000\000\000\000\000\000\000\370.
Jun 29 20:00:43.749 INFO [region 2] 3 epoch changed conf_ver: 1 version: 10 != conf_ver: 1 version: 9, retry later
Jun 29 20:00:43.850 INFO [region 2] on split region at key t\200\000\000\000\000\000\000\377\027\000\000\000\000\000\000\000\370.
Jun 29 20:00:43.850 DEBG scheduling task ask split region 2 with key t\200\000\000\000\000\000\000\377\027\000\000\000\000\000\000\000\370
Jun 29 20:00:43.850 DEBG loop poll - 145.552154ms
Jun 29 20:00:43.851 DEBG loop time - Instant { t: 415878455685656 }
Jun 29 20:00:43.851 DEBG loop process - 1 events, 18.385µs
Jun 29 20:00:43.851 DEBG executing task ask split region 2 with key t\200\000\000\000\000\000\000\377\027\000\000\000\000\000\000\000\370
Jun 29 20:00:43.851 DEBG reconnect remains: 10
Jun 29 20:00:43.851 DEBG loop poll - 7.582µs
Jun 29 20:00:43.851 DEBG loop time - Instant { t: 415878455772221 }
Jun 29 20:00:43.851 DEBG request sent: 1
Jun 29 20:00:43.851 DEBG loop process - 1 events, 151.553µs
Jun 29 20:00:43.852 DEBG loop poll - 9.057µs
Jun 29 20:00:43.852 DEBG loop time - Instant { t: 415878455943688 }
Jun 29 20:00:43.852 DEBG loop process - 0 events, 11.349µs
Jun 29 20:00:43.852 DEBG loop poll - 221.894µs
```

Note: source like `pd.rs:437` is missing. This will lead to notable inconvenience when tracing issues.

### PR DESCRIPTION
**Pull request by Hoverbear**
## What have you changed? (mandatory)

Adds a custom log formatter which includes log location and improves date format.

Fixes #3257 .
Fixes #3261 .

After:

```
2018/06/29 16:39:21.518 WARN src/bin/tikv-server.rs:87: Limit("kernel parameters net.core.somaxconn got 128, expect 32768")
2018/06/29 16:39:21.518 WARN src/bin/tikv-server.rs:87: Limit("kernel parameters net.ipv4.tcp_syncookies got 1, expect 0")
2018/06/29 16:39:21.519 WARN src/bin/tikv-server.rs:87: Limit("kernel parameters vm.swappiness got 60, expect 0")
2018/06/29 16:39:21.519 WARN src/bin/tikv-server.rs:92: environment variable `TZ` is missing, using `/etc/localtime`
2018/06/29 16:39:21.519 INFO src/util/config.rs:820: data_path: "/tmp/kv/1", mount fs info: FsInfo { tp: "tmpfs", opts: "rw,nosuid,nodev", mnt_dir: "/tmp", fsname: "tmpfs" }
2018/06/29 16:39:21.519 WARN src/bin/tikv-server.rs:97: FileSystem("fs: \"tmpfs\" no device find in block")
2018/06/29 16:39:21.519 INFO src/util/config.rs:820: data_path: "/tmp/kv/1/raft", mount fs info: FsInfo { tp: "tmpfs", opts: "rw,nosuid,nodev", mnt_dir: "/tmp", fsname: "tmpfs" }
2018/06/29 16:39:21.519 WARN src/bin/tikv-server.rs:101: FileSystem("fs: \"tmpfs\" no device find in block")
2018/06/29 16:39:21.519 INFO src/pd/util.rs:391: connect to PD endpoint: "127.0.0.1:2379"
2018/06/29 16:39:21.520 ERRO src/pd/util.rs:351: PD endpoint 127.0.0.1:2379 failed to respond: Grpc(RpcFailure(RpcStatus { status: Unavailable, details: Some("Connect Failed") }))
```

## What are the type of the changes? (mandatory)

- Improvement (non-breaking change which is an improvement to an existing feature)

## How has this PR been tested? (mandatory)

Manually

## Does this PR affect documentation (docs/docs-cn) update? (mandatory)

Nope

## Does this PR affect tidb-ansible update? (mandatory)
Nope

### CODE REVIEW

**Comment by Hoverbear** [comment]
@breeswish I know this isn't the original format, but could you PTAL? We can use the original format if it's better, but I think this is maybe more clear.

**Comment by siddontang** [comment]
can we remove `src/bin` or other parent paths? I worry the path may be long sometimes.

**Comment by breeswish** [comment]
@Hoverbear Yes, I understand, this is less ambiguous than YYYY/MM/DD for international users, but for historical reasons we should keep it, otherwise it will be tough for us to "grep" something among different components when solving clients' issues.

**Comment by Hoverbear** [comment]
@breeswish Oh, I meant regarding how the file locations are after the log level. :)

**Comment by siddontang** [comment]
@breeswish 

> this is less ambiguous than YYYY/MM/DD for international users

What do you mean?

**Comment by Hoverbear** [comment]
@siddontang We can, we can also do `record.module()` if we just want to module instead of a file name.

**Comment by Hoverbear** [comment]
Example now: 

```
2018/07/03 12:03:59.564 WARN tikv-server.rs:87: Limit("kernel parameters net.core.somaxconn got 128, expect 32768")
2018/07/03 12:03:59.564 WARN tikv-server.rs:87: Limit("kernel parameters net.ipv4.tcp_syncookies got 1, expect 0")
2018/07/03 12:03:59.564 WARN tikv-server.rs:87: Limit("kernel parameters vm.swappiness got 60, expect 0")
2018/07/03 12:03:59.565 WARN tikv-server.rs:92: environment variable `TZ` is missing, using `/etc/localtime`
2018/07/03 12:03:59.565 INFO config.rs:820: data_path: "/tmp/kv/1", mount fs info: FsInfo { tp: "tmpfs", opts: "rw,nosuid,nodev", mnt_dir: "/tmp", fsname: "tmpfs" }
2018/07/03 12:03:59.565 WARN tikv-server.rs:97: FileSystem("fs: \"tmpfs\" no device find in block")
2018/07/03 12:03:59.565 INFO config.rs:820: data_path: "/tmp/kv/1/raft", mount fs info: FsInfo { tp: "tmpfs", opts: "rw,nosuid,nodev", mnt_dir: "/tmp", fsname: "tmpfs" }
2018/07/03 12:03:59.565 WARN tikv-server.rs:101: FileSystem("fs: \"tmpfs\" no device find in block")
2018/07/03 12:03:59.566 INFO util.rs:391: connect to PD endpoint: "127.0.0.1:2379"
2018/07/03 12:03:59.566 ERRO util.rs:351: PD endpoint 127.0.0.1:2379 failed to respond: Grpc(RpcFailure(RpcStatus { status: Unavailable, details: Some("Connect Failed") }))
2018/07/03 12:03:59.566 ERRO tikv-server.rs:418: failed to create rpc client: Other(StringError("[src/pd/util.rs:382]: PD cluster failed to respond"))
```

**Comment by breeswish** [comment]
@siddontang User might be confused with YYYY/MM/DD or YYYY/DD/MM

**Comment by breeswish** [comment]
Tested manually, seems good! LGTM

**Comment by Hoverbear** [comment]
@breeswish I'm not sure I've ever seen `YYYY/DD/MM` :thinking:

**Comment by siddontang** [comment]
PTAL @overvenus

**Comment by siddontang** [comment]
CI failed @Hoverbear

File: src/util/logger.rs [hunk 1]
```diff
@@ -97,3 +103,320 @@ fn test_get_level_by_string() {
...
+    use slog::Logger;
+    use slog_term::PlainSyncDecorator;
+    use std::cell::RefCell;
+    use std::io::Write;
+    use std::str::from_utf8;
+
+    // Due to the requirements of `Logger::root*` on a writer with a 'static lifetime
+    // we need to make a Thread Local,
+    // and implement a custom writer.
+    thread_local! {
```
Comment (overvenus) [inline src/util/logger.rs]: How about `lazy_static!`? I worry about context switch may fail the test.

File: src/util/logger.rs [hunk 2]
```diff
@@ -97,3 +103,320 @@ fn test_get_level_by_string() {
...
+            self.count += buf.len();
+            ()
+        })
+    }
+}
+
+struct Serializer<'a> {
+    comma_needed: bool,
+    decorator: &'a mut RecordDecorator,
+    stack: Vec<(String, String)>,
```
Comment (overvenus) [inline src/util/logger.rs]: Where do we push things into the stack?

File: src/util/logger.rs (hunk 1 above)
Comment (breeswish) [reply src/util/logger.rs]: Context switch should not change thread local variables.

Comment (overvenus) [reply src/util/logger.rs]: What if context switch happens between L398 nad L409?

Comment (breeswish) [reply src/util/logger.rs]: I can't see the problem. Could you illustrate the detail of your concern?

Comment (overvenus) [reply src/util/logger.rs]: You are right, there's nothing to worry about!

File: src/util/logger.rs (hunk 2 above)
Comment (Hoverbear) [reply src/util/logger.rs]: You're right! removed!

**Comment by siddontang** [comment]
Ping @breeswish @overvenus

**Comment by siddontang** [comment]
Ping @overvenus

**Comment by Hoverbear** [comment]
@overvenus I had to fix rustfmt, can you please take a look again?

**Pull request merged**